### DISCUSSION TITLE
[FEATURE] Immich server logs proxy IP instead of client IP when Login failed

### DISCUSSION THREAD

**Discussion by babushanaelsa** [immich-app/immich#1602] [Feature Request]
**Describe the bug**
When I try to login to my Immich and use wrong credentials, the server log shows just the IP of the Proxy. This happens in both cases:

1. Using the provided Immich_proxy which is setup if using docker compose
2.  If I replace the provided Proxy with my "main" NGINX that I use for all the other services

The necessary proxy headers seem to be correct.
If I forward the exact same Proxy Host (without changing config) to another service and create a failed Login the correct Client IP is logged.

Google taught me, that an application server also needs to have some kind of headers to accept / read the transmitted client IP, but I'm not experianced enough to see if this is the case with Immich-Server or not.

**Task List**

*Please complete the task list below. We need this information to help us reproduce the bug or point out problems in your setup. You are not providing enough info may delay our effort to help you.*

- [x ] I have read thoroughly the README setup and installation instructions.
- [ ] I have included my `docker-compose` file.
- [ ] I have included my redacted `.env` file.
- [ x] I have included information on my machine, and environment.

**To Reproduce**
1. Go to Immich Login page
2. Put in fantasy login credentials
4. Click on Login
5. Wait for error message "wrong credentials"
6. Check immich server log for the provided IP adress in the authentication error line

**Expected behavior**
Immich should log the Clients IP, in this case to setup a correct fail2ban protection. If somebody tries to hack your immich, it doesn't block the whole access because the proxy is blocked, it just blocks the IP of the guy trying to get access

**Screenshots**

**System**

- Server Version: `latest Release-Tag on docker hub`

**Additional context**

**Comment by raisinbear** [comment]
@babushanaelsa in the mean time you could use the reverse proxy access logs to build a fail2ban filter. I haven't had a look what the server itself logs, but I'm using (my own, not the included) nginx to proxy to it and a failed login attempt will produce a line in the _web server access_ log similar to
`[CLIENT IP] - - [19/Jan/2023:06:35:54 +0100] "POST /api/auth/login HTTP/2.0" 400 80 "[URL]" "[USER AGENT]"`

I hardly have any experience with fail2ban filter creation, let alone regex. But I cobbled together this criterion
`failregex = ^<HOST> -.*"POST.*/api/auth/login.*HTTP.*" 400`
which has been working fine for some time now and will correctly block the tresspassing IP. I haven't really investigated the chance of false positives, but remember seeing them with the mobile app on occasion. Don't remember exactly what the circumstances, however. Might have been logged out for some reason. Since I only block on the 3rd try, however, I never locked anyone with valid credentials out so far ;)

**Comment by babushanaelsa** [comment]
@raisinbear 

Thx for your feedback, I do use a fail2ban filter already and it works. But I think the proper way would not be to "shut down" the whole service in worst case but to block the suspicious attempt 🙂 Beyond that you can also establish other protection layers like geoblocking, but therefore the correct IP adress is needed.

**Comment by raisinbear** [comment]
@babushanaelsa
That's what I was trying to say: You get the correct IP, not proxy but client. But in the (nginx in my case) server log, (apparently) not the immich server log. So you can specifically block the violating IP, not the proxy which would shut down your service. Just tried again and it works - as it should, because the IP logged is collected by nginx not immich. Nginx merely links the client IP to immich's response in the log.

**Comment by jon-f-novabackup** [comment]
I'm having trouble with @rasinbear's posted failregex statement in the 2nd post in this thread, to get it to pass without seeing **'ERROR: No failure-id group in '/home/jon/appdata/swag/fail2ban/filter.d/immich-unauthorized.conf'** after I manually issue the 'docker exec -it swag fail2ban-regex /home/jon/appdata/swag/log/nginx/access.log /home/jon/appdata/swag/fail2ban/filter.d/immich-unauthorized.conf' command to check if the syntax of the regex is valid and that it finds a result in my nginx access log; if someone could help me out that would be great, and I will try to include all necessary info for what my config is and what I'm expecting to get working, to be able to ban invalid Immich website logins, which by default fail2ban and nginx won't do that, as it takes some manual work to get it to know how. I have docker-immich running under nginx (that is part of the docker-swag container, that includes nginx and fail2ban) and I want to have invalid logins to Immich web server to trigger the banning of the IP address after X invalid login attempts. I know what the fail2ban that docker-swag has running as a container is, for reasons outside of Immich website login dialog, able to ban the same IP address that I used in the example log statements below, because I see it show those banned IPs in the fail2ban jail status command, in the 'nginx-unauthorized' jail, but it just will NEVER ban invalid logins done by Immich website dialog 'https://immichexample.ddns.net:36800/auth/login' (Note: that is not the real website address and the IP address is not real, I have remarked that in this post). I'm running the latest version of nginx and fail2ban (in the latest docker-swag container) and the latest Immich 1.54.1 server version. I tried to utilize the regex statement for Immich provided by @raisinbear in the 2nd post in this discussion thread, which was this "failregex" statement:

`failregex = ^<HOST> -.*"POST.*/api/auth/login.*HTTP.*" 400`

When I setup my filter .conf file in fail2ban for Immich unauthorized logins when I go to test it manually with the '**docker exec -it swag fail2ban-regex /home/jon/appdata/swag/log/nginx/access.log /home/jon/appdata/swag/fail2ban/filter.d/immich-unauthorized.conf'** command it either can't match any values to be able to add a ban or it shows this error: **'ERROR: No failure-id group in '/home/jon/appdata/swag/fail2ban/filter.d/immich-unauthorized.conf'**. I take it that it has a problem with my syntax in the actual .conf script below but at this point I have no idea why it is erroring when I test it with the same 'fail2ban-regex' command stated above. I haven't tried implementing it in my 'jail.local' file yet to add that new non-stock filter there for that reason, because the output with the fail2ban-regex command is showing that failure with nothing getting matched in the output of the above command. Is it possible that the syntax is correct here and that I just need to try to add that new non-stock 'immich-unauthorized' filter into the 'jail.local' file now and not worry about the regex testing error that I keep seeing mentioned above?

The contents of my '/home/jon/appdata/swag/fail2ban/filters.d/immich-unauthorized.conf' file are this:

[Definition]
`failregex = ^<HOST> -.*"POST.*/api/auth/login.*HTTP.*" 400`

The contents of my '/home/jon/appdata/swag/fail2ban/jail.local' file, that exists in the default path that docker-swag installs config files in (in my home folder), for fail2ban, shows this content (notice I have remarked out the #ignoreip section so it temporarily doesn't utilize that to make banning testing easier to notice) are this:

jon@pi4:~/appdata/swag/fail2ban$ cat jail.local
[DEFAULT]
ignoreip =
banaction = iptables-allports
bantime = 600
findtime = 600

maxretry = 5

[ssh]
enabled = false

[nginx-http-auth]
enabled = true
filter = nginx-http-auth
port = http,https
logpath = /config/log/nginx/error.log

[nginx-badbots]
enabled = true
port = http,https
filter = nginx-badbots
logpath = /config/log/nginx/access.log
maxretry = 2

[nginx-botsearch]
enabled = true
port = http,https
filter = nginx-botsearch
logpath = /config/log/nginx/access.log

[nginx-deny]
enabled = true
port = http,https
filter = nginx-deny
logpath = /config/log/nginx/error.log

[nginx-unauthorized]
enabled = true
port = http,https
filter = nginx-unauthorized
logpath = /config/log/nginx/access.log

But I get this result "ERROR: No failure-id group", which I gather is a common error when attempting to filter logs with fail2ban. I still can't get fail2ban to ban any IP address that I'm attempting invalid logins with (and like I said I removed the default ignoreip entries), and could use some help. I'm just using a stock install of the immich docker and swag docker, which comes with nginx and fail2ban, and haven't made any modifications there. The IP address or IP range that I'm utilizing to attack the login page with invalid logins is not part of the IP exclusions inside jail.conf or any other config file that I'm aware of. Can someone assist me in getting fail2ban to detect and ban Immich website (running under nginx reverse proxy) invalid logins. I have Immich 1.54.1 server running in the standard docker-compose / docker image, and SWAG docker is my reverse proxy (to get HTTPS working for Immich) but I know that it is not banning based on invalid login attempts on the web login dialog. I need it to be able to detect invalid logins with Immich 1.54.1 server running in the standard docker-install container, where nginx reverse proxy does show log entries in my swag config folder, where nginx logs store to, which is how swag works by default, (/home/jon/appdata/swag/log/nginx/access.log) like mentioned below (the IP address seen at the start is the IP address I'm doing the invalid logins on, and in that case it is not part of "ignoreip" section of any fail2ban config file that I know of, so it should get banned from there) when the login is invalid (at least I believe that is the case, that "400 80" in the access.log statement equates to an invalid login when I tail the log and try invalid logins on the Immich web login dialog):

nginx based access.log errors appear like this for invalid logins to Immich web server 1.54.1:
68.15.200.50 - - [24/Apr/2023:10:45:07 -0700] "POST /api/auth/login HTTP/1.1" 400 80 "https://immichexample.ddns.net:36800/auth/login" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/112.0.0.0 Safari/537.36"

And if I run 'docker logs immich_server' to tail the docker container for immich_server after the invalid login attempts that I attempt via the Immich website (behind nginx reverse proxy) I see the invalid login attempts in the 'docker logs immich_server' command output like this:

'docker logs immich_server -f' tail shows a bunch of failed login attempts to immich_web_server, as I only have one website hosted on this system where I would be utilizing this email address for the invalid login testing, which was my Immich website login dialog:
**[Nest] 1  - 04/24/2023, 5:50:18 PM    WARN [AuthService] Failed login attempt for user test@gmail.com from ip address 68.15.200.50**

There does not appear to be a log file produced by the running 'immich_server' container, that would store to the local file system it seems, so maybe that is part of some additional setup I need to do in order to get that immich_server log output to exist in a local file that fail2ban can use for input to read, in any case I copied the text from the output of the prior command to a local .txt file and am using that for the input log file for 'docker exec -it swag fail2ban-regex' command but everytime I try to setup a filter.d .conf file or try manually with 'docker exec -it swag fail2ban-regex' command it either can't match any values to be able to add a ban or it shows this error: "ERROR: No failure-id group in '/home/jon/appdata/swag/fail2ban/filter.d/immich-unauthorized.conf'". 

I have tried using this filter conf example from fail2ban's official github repo:
https://github.com/fail2ban/fail2ban/blob/master/config/filter.d/nginx-bad-request.conf

and this filter conf which comes with docker-swag:
https://github.com/linuxserver/docker-swag/blob/master/root/defaults/fail2ban/filter.d/nginx-unauthorized.conf

I'm trying to use the above examples and so far when attempting to utilize them stock or after modifications it keeps showing "ERROR: No failure-id group" testing with the 'docker exec -it swag fail2ban-regex' command.

I could use some assistance here if possible! Note: My computer's date and time is valid, which I read can cause problems with fail2ban if not accurate. When the failed logins are attempted on Immich web server auth/login dialog (not using OAuth/2FA in Immich) /home/jon/appdata/swag/log/nginx/access.log entries are created but not /home/jon/appdata/swag/log/nginx/server.log entries, when it comes to nginx, and I gather that is normal.

**Comment by simone-viozzi** [comment]
Hi, 
If the end goal is to create a fail2ban config, you can follow https://github.com/immich-app/immich/discussions/3243#discussioncomment-6681948.

In my configuration I use traefik as a reverse proxy, but it should not be relevant since I use the logs of `immich-server` to create the fail2ban config.

**Comment by jrasm91** [comment]
Is this still an issue? We've added support for proxy headers and fixed a bug with logging the client id.

If you follow the [instructions here](https://immich.app/docs/administration/reverse-proxy) immich should correctly log the client IP.

**Reply by Snyaify** [reply]
What's about traefik example config as reverse proxy?
